backend/windmill-worker/src/worker_flow.rs

```diff
@@ -1536,11 +1557,63 @@ async fn compute_skip_branchall_failure<'c>(
 //         )))
 // }
 
-fn next_retry(retry: &Retry, status: &RetryStatus) -> Option<(u32, Duration)> {
-    (status.fail_count <= MAX_RETRY_ATTEMPTS)
-        .then(|| &retry)
-        .and_then(|retry| retry.interval(status.fail_count, false))
-        .map(|d| (status.fail_count + 1, std::cmp::min(d, MAX_RETRY_INTERVAL)))
+async fn evaluate_retry(
+    retry: &Retry,
+    status: &RetryStatus,
+    result: Arc<Box<RawValue>>,
+    flow_args: Option<Marc<HashMap<String, Box<RawValue>>>>,
+    flow_id: Option<uuid::Uuid>,
+    client: Option<&AuthedClient>,
+    db: Option<&DB>,
+) -> anyhow::Result<Option<(u32, Duration)>> {
+    if status.fail_count > MAX_RETRY_ATTEMPTS {
+        return Ok(None);
+    }
+
+    if let Some(retry_if) = &retry.retry_if {
+        let args = match flow_args {
+            Some(args) => args,
+            None => {
+                let flow_id = flow_id.ok_or_else(|| {
+                    Error::internal_err(
+                        "flow_id required when flow_args not provided for retry condition"
+                            .to_string(),
+                    )
+                })?;
+                let db = db.ok_or_else(|| {
+                    Error::internal_err(
+                        "database required when flow_args not provided for retry condition"
+                            .to_string(),
+                    )
+                })?;
+
+                let fetched_args = sqlx::query_scalar!(
+                    "SELECT args AS \"args: Json<HashMap<String, Box<RawValue>>>\" FROM v2_job WHERE id = $1",
+                    flow_id
+                )
+                .fetch_one(db)
+                .await
+                .map_err(|e| {
+                    Error::internal_err(format!("retrieval of args from state for retry condition: {e:#}"))
+                })?;
```

Comment:
are you sure the flow args are not already available and fetched before and can't be reused?